File: csharp/IceBox/Greeter/Service/GreeterService.cs
// Copyright (c) ZeroC, Inc.

using System.Diagnostics;
using VisitorCenter;

namespace Service;

/// <summary>An IceBox service that provides an implementation of the Greeter interface.</summary>
internal class GreeterService : IceBox.Service
{
    // The object adapter created by this service.
    private Ice.ObjectAdapter? _adapter;

    /// <inheritdoc/>
    public void start(string name, Ice.Communicator communicator, string[] args)
    {
        // Create an object adapter that listens for incoming requests and dispatches them to servants.
        Debug.Assert(_adapter is null);
        _adapter = communicator.createObjectAdapterWithEndpoints("GreeterAdapter", "tcp -p 4061");

        // Register the Chatbot servant with the adapter.
        _adapter.add(new Chatbot("Syd"), Ice.Util.stringToIdentity("greeter"));

        // Start dispatching requests.
        _adapter.activate();
        Console.WriteLine("Listening on port 4061...");
    }

    /// <inheritdoc/>
    public void stop()
    {
        Console.WriteLine("Shutting down...");

        // We destroy the object adapter in stop() in case the service is started, stopped and then restarted again
        // programmatically or using an admin tool.
        // The general rule is stop should cleanup all resources created by start.
        Debug.Assert(_adapter is not null);
        _adapter.destroy();
        _adapter = null;
    }
}

File: java/IceBox/greeter/service/src/main/java/com/example/icebox/greeter/service/GreeterService.java
// Copyright (c) ZeroC, Inc.

package com.example.icebox.greeter.service;

import com.zeroc.Ice.Communicator;
import com.zeroc.Ice.Identity;
import com.zeroc.Ice.ObjectAdapter;
import com.zeroc.IceBox.Service;

/**
 * An IceBox service that provides an implementation of the Greeter interface.
 */
public class GreeterService implements Service {

    // The object adapter created by this service.
    private ObjectAdapter _adapter;

    @Override
    public void start(String name, Communicator communicator, String[] args) {
        // Create an object adapter that listens for incoming requests and dispatches them to servants.
        _adapter = communicator.createObjectAdapterWithEndpoints("GreeterAdapter", "tcp -p 4061");

        // Register the Chatbot servant with the adapter.
        _adapter.add(new Chatbot(), new Identity("greeter", ""));

        // Start dispatching requests.
        _adapter.activate();
        System.out.println("Listening on port 4061...");
    }

    @Override
    public void stop() {
        System.out.println("Shutting down...");

        // We destroy the object adapter in stop() in case the service is started, stopped and then restarted again
        // programmatically or using an admin tool.
        // The general rule is stop should cleanup all resources created by start.
        assert _adapter != null;
        _adapter.destroy();
        _adapter = null;
    }
}
